modules/custom/falcon_mail/src/Plugin/Mail/FalconMailSystem.php

```diff
@@ -0,0 +1,138 @@
+      $message['body'] = implode("\n\n", $message['body']);
+    }
+
+    // Wrap body in theme template.
+    if (!empty($message['params']['theme_template'])) {
+      $message['body'] = str_replace('#$#BODY#$#', $message['body'], $message['params']['theme_template']);
+    }
+
+    // Replace tokens.
+    if (!empty($message['params']['replace_tokens'])) {
+
+      $tokens = !empty($message['params']['render_tokens']) ? $message['params']['render_tokens'] : [];
+      $token_options = !empty($message['params']['token_options']) ? $message['params']['token_options'] : [];
+
+      // Replace tokens and set new body.
+      $message['body'] = Drupal::token()->replace($message['body'], $tokens, $token_options);
+    }
+
+    // Replace relative urls in body.
+    $message['body'] = $this->replaseRelativeUrlsWithSaveStyles($message['body']);
+
+    return $message;
+  }
+
+  /**
+   * Send the e-mail message with using default PHP_Mail.
+   *
+   * @param array $message
+   *   A message array.
+   *
+   * @return bool
+   *   TRUE if the mail was successfully accepted, otherwise FALSE.
+   */
+  public function mail(array $message) {
+    return parent::mail($message);
+  }
+
+  /**
+   * Replace relative urls on absolute with saving <html>, <body>, <head> tags.
+   *
+   * @param string $body
+   *   Mail body.
+   *
+   * @return string
+   *   String with/without replaced relative urls.
+   */
+  public function replaseRelativeUrlsWithSaveStyles($body) {
+    try {
+      // Create Global dom document.
+      $html = new \DOMDocument();
+      $html->loadHTML($body);
+
+      // Get body element.
+      $bodyElement = $html->getElementsByTagName('body')[0];
+
+      if (empty($bodyElement)) {
+        return Html::transformRootRelativeUrlsToAbsolute((string) $body, \Drupal::request()
+          ->getSchemeAndHttpHost());
+      }
+
+      // Get body structure as string.
+      $strBody = $html->saveHTML($html->getElementsByTagName('body')[0]);
+
+      // Replace all relative urls in body.
+      $strNewBody = Html::transformRootRelativeUrlsToAbsolute((string) $strBody, \Drupal::request()
+        ->getSchemeAndHttpHost());
+
+      // Create new HTML document with new body.
+      $newBodyHtml = Html::load($strNewBody);
+      // Get new body element.
+      $newBodyElement = $newBodyHtml->getElementsByTagName('body')[0];
+      // Provide new body element to Global dom document.
+      $newBodyElement = $html->importNode($newBodyElement, TRUE);
+
+      // Remove all childNodes from body.
+      while ($bodyElement->childNodes[0]) {
+        $bodyElement->removeChild($bodyElement->childNodes[0]);
+      }
+      // Add childNodes from new body.
+      while ($newBodyElement->childNodes[0]) {
+        $bodyElement->appendChild($newBodyElement->childNodes[0]);
+      }
+
+      // Save Global dom document to message body.
+      $body = $html->saveHTML();
+    }
+    catch (Exception $e) {
+      Drupal::logger('falcon_mail')
```

Comment:
I'd use watch `watchdog_exception()` to log exception details.

Reply:
Changed.